### PR TITLE
Make Indexer classes not inherit from tuple.

### PR DESCRIPTION
**Pull request by shoyer**
I'm not entirely sure this is a good idea. The advantage is that it ensures that
all our indexing code is entirely explicit: everything that reaches a backend
*must* be an ExplicitIndexer. The downside is that it removes a bit of
internal flexibility: we can't just use tuples in place of basic indexers
anymore. On the whole, I think this is probably worth it but I would appreciate
feedback.

@fujiisoup can you take a look?

 - [x] Tests added / passed
 - [x] Passes ``git diff upstream/master **/*py | flake8 --diff``

### CODE REVIEW

**Comment by fujiisoup** [comment]
This looks pretty clean and less error-prone.

For more cleanliness, I'm wondering if we could more clearly distinguish between raw array-wrappers (such as `NumpyIndexingAdapter`) and wrappers of array-wrapper (such as `MemoryCachedArray`).
But as a whole, I like this idea.

Regarding the more array-type support in the future (as suggested in [comment](https://github.com/pydata/xarray/issues/1617#issuecomment-340290288)),
is there something to prepare in this PR?
I guess there are some typical indexing types, such as Fortran-type and Numpy-type.
Can we have some abstract classes (maybe too early)?

**Comment by shoyer** [comment]
There's been discussion about abstract classes on the NumPy mailing list.
It would be nice to use something standard if possible:
https://mail.python.org/pipermail/numpy-discussion/2017-November/thread.html#77312

File: xarray/core/variable.py [hunk 1]
```diff
@@ -766,6 +769,7 @@ def chunk(self, chunks=None, name=None, lock=False):
             if utils.is_dict_like(chunks):
                 chunks = tuple(chunks.get(n, s)
                                for n, s in enumerate(self.shape))
+            data = indexing.WrapImplicitIndexing(data, indexing.OuterIndexer)
```
Comment (fujiisoup) [inline xarray/core/variable.py]: Why is this line necessary here? (and only OuterIndexer?)
I think it would be better if we could have a simple comment above.

File: xarray/core/indexing.py [hunk 2]
```diff
@@ -277,84 +278,224 @@ def _index_indexer_1d(old_indexer, applied_indexer, size):
...
+            VectorizedIndexer: 'Vectorized'
+        }[type(key)]
+        raise NotImplementedError(
+            '{} indexing for {} is not implemented. Load your data first with '
+            '.load(), .compute() or .persist(), or disable caching by setting '
+            'cache=False in open_dataset.'.format(key_type_name, type(target)))
+    return key.tuple
+
+
+class WrapImplicitIndexing(utils.NDArrayMixin):
```
Comment (fujiisoup) [inline xarray/core/indexing.py]: This name sounds like a function which behaves opposite to `unwrap_explicit_indexer`. Can we change this something like `ImplicitIndexedArrayWrapper`?
(But as the below comment, I did not yet understand how this wrapper will be used.)

File: xarray/core/indexing.py [hunk 3]
```diff
@@ -277,84 +278,224 @@ def _index_indexer_1d(old_indexer, applied_indexer, size):
...
+        super(VectorizedIndexer, self).__init__(new_key)
+
+
+class ExplicitlyIndexed(object):
+    """Mixin to mark support for Indexer subclasses in indexing."""
 
 
-class VectorizedIndexer(IndexerTuple):
-    """ Tuple for vectorized indexing """
+class ExplicitNDArrayMixin(utils.NDArrayMixin, ExplicitlyIndexed):
```
Comment (fujiisoup) [inline xarray/core/indexing.py]: I like a more explicit name such as `ExplicitlyIndexedNDArrayMixin`. Is it too long?

File: xarray/core/indexing.py [hunk 4]
```diff
@@ -277,84 +278,224 @@ def _index_indexer_1d(old_indexer, applied_indexer, size):
...
+    indexed with an integer are dropped from the result.
     """
+    def __init__(self, key):
+        if not isinstance(key, tuple):
+            raise TypeError('key must be a tuple: {!r}'.format(key))
+
+        new_key = []
+        for k in key:
+            if isinstance(k, integer_types):
+                k = int(k)
```
Comment (fujiisoup) [inline xarray/core/indexing.py]: This looks a good idea over checking an integer type in array wrappers :+1:

File: xarray/core/variable.py (hunk 1 above)
Comment (shoyer) [reply xarray/core/variable.py]: Added a comment to explain, with a link to the upstream dask issue (https://github.com/dask/dask/issues/2883).

**Comment by shoyer** [comment]
I plan to merge this shortly and issue another release candidate unless any one has objections.

It does occur to me that this could break custom backends, since they will not longer be getting tuples as indexers. @rabernat any ideas for how to minimize that pain?

File: xarray/core/variable.py [hunk 5]
```diff
@@ -1261,7 +1270,7 @@ def equals(self, other, equiv=duck_array_ops.array_equiv):
             return (self.dims == other.dims and
                     (self._data is other._data or
                      equiv(self.data, other.data)))
-        except (TypeError, AttributeError):
+        except (TypeError, AttributeError) as e:
```
Comment (jhamman) [inline xarray/core/variable.py]: it doesn't look like `e` is ever used.

Comment (shoyer) [reply xarray/core/variable.py]: Good catch, this was introduced in debugging

**Comment by shoyer** [comment]
Actually, looking over @rabernat's custom backend code in xmitgcm, I think that should be OK. This would only be a problem if you pass custom duck arrays into `xarray.Variable` objects, rather than numpy or dask arrays.

**Comment by shoyer** [comment]
OK, merging. Hopefully we'll catch any issues in the next release candidate!
